### PR TITLE
Add needed pkgs, classes, and interfaces

### PR DESCRIPTION
**Pull request by Mmduh-483**
Define required components of the project.
config: Read the configurations.
daemon: daemon define the main service for the project.
guid: manage the guids pool.
k8s-client: interface for kubernetes client and its releated need methods.
sm: subnet manger pkg, where subnet manger client is loaded as plugin.
watcher: define kubernetes events and watcher to handle the events.

### CODE REVIEW

File: pkg/types/types.go [hunk 1]
```diff
@@ -1,4 +1,58 @@
...
+	// It returns error if failed to create the pool.
+	GeneratePool() error
+
+	// NextGUID return the next available guid in the pool.
+	// It returns error if there is no more available guids.
+	NextGUID() error
+}
+
+type SubnetMangerClient interface {
+	// Name returns the subnet manger name.
```
Comment (moshe010) [inline pkg/types/types.go]: s/manger/manager

File: pkg/types/types.go [hunk 2]
```diff
@@ -1,4 +1,58 @@
...
+	// GeneratePool return the next available guid in the pool.
+	// It returns error if failed to create the pool.
+	GeneratePool() error
+
+	// NextGUID return the next available guid in the pool.
+	// It returns error if there is no more available guids.
+	NextGUID() error
+}
+
+type SubnetMangerClient interface {
```
Comment (moshe010) [inline pkg/types/types.go]: s/SubnetMangerClient/SubnetManagerClient

File: pkg/types/types.go [hunk 3]
```diff
@@ -1,4 +1,58 @@
...
+	AddPodHandler(interface{}) error
+
+	// DeletePodHandler handles pod deleting event, for each deleted pod.
+	// Check pod's annotations to know if pod is InfiniBand app to remove its pkey and guid.
+	// It returns error in case of failure.
+	DeletePodHandler(interface{}) error
+}
+
+type GuidPool interface {
+	// StartRange returns the first guid in the pool
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/types/types.go]: It is possible to save it to a file, or do as kubemacpool, when the manager start or restart it reads all pods, then reads the mac addresses reserved for them.
https://github.com/k8snetworkplumbingwg/kubemacpool/blob/master/pkg/pool-manager/pod_pool.go#L194, we can use this approach to keep tracking available and reserved  GUID.

What do you think?

File: pkg/types/types.go [hunk 4]
```diff
@@ -1,4 +1,58 @@
...
+	StartRange() string
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/types/types.go]: Start and End range will be configured in a config map

File: pkg/types/types.go (hunk 2 above)
Comment (Mmduh-483) [reply pkg/types/types.go]: fixed

File: pkg/types/types.go (hunk 1 above)
Comment (Mmduh-483) [reply pkg/types/types.go]: fixed

File: pkg/types/types.go [hunk 5]
```diff
@@ -1,4 +1,58 @@
 package types
 
 type Daemon interface {
+	// Init initializes the need components including k8s client, subnet manger client plugins, and guid pool.
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/types/types.go]: moved

File: pkg/types/types.go [hunk 6]
```diff
@@ -1,4 +1,58 @@
...
+	// It returns error in case of failure.
+	AddPodHandler(interface{}) error
+
+	// DeletePodHandler handles pod deleting event, for each deleted pod.
+	// Check pod's annotations to know if pod is InfiniBand app to remove its pkey and guid.
+	// It returns error in case of failure.
+	DeletePodHandler(interface{}) error
+}
+
+type GuidPool interface {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/types/types.go]: moved

File: pkg/types/types.go [hunk 7]
```diff
@@ -1,4 +1,58 @@
...
+	// AddPKey add pkey for the given guid.
+	// It return error if failed to add the pkey for specified guid.
+	AddPKey(pkey, guid string) error
+
+	// RemovePKey remove pkey for the given guid.
+	// It return error if failed to remove the pkey for specified guid.
+	RemovePKey(pkey, guid string) error
+}
+
+type PluginLoader interface {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/types/types.go]: moved

File: pkg/types/types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/types/types.go]: moved

File: pkg/sm/plugins/types.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
+package plugins
```
Comment (moshe010) [inline pkg/sm/plugins/types.go]: change file for plugin.go

File: pkg/guid/guid_pool.go [hunk 9]
```diff
@@ -0,0 +1,18 @@
+package guid
+
+type GuidPool interface {
+	// AllocateGUID allocate the next free guid in the range.
+	// It returns the allocated guid or error if range is full.
+	AllocateGUID() (string, error)
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/guid/guid_pool.go]: I missed it, added the function `ReleaseGUID`

File: pkg/sm/plugins/types.go (hunk 8 above)
Comment (Mmduh-483) [reply pkg/sm/plugins/types.go]: changed the file name to plugin.go

**Comment by Mmduh-483** [comment]
@moshe010 updated the PR.
the build is going to fail due to missed vendor packages.

File: pkg/k8s-client/client.go [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+	"time"
+
+	kapi "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+)
+
+type Client interface {
+	GetPod(namespace, name string) (*kapi.Pod, error)
```
Comment (moshe010) [inline pkg/k8s-client/client.go]: not sure we will need all the method, but I don't want to block it

**Pull request merged**